include/swift/IDE/SourceEntityWalker.h

```diff
@@ -82,11 +82,26 @@ class SourceEntityWalker {
 
   /// This method is called when first visiting a decl, before walking into its
   /// children.  If it returns false, the subtree is skipped.
-  virtual bool walkToDeclPre(Decl *D, CharSourceRange Range) { return true; }
+  virtual bool walkToDeclPre(Decl *D, CharSourceRange Range) {
+    beginBalancedDeclVisit(D);
+    return true;
+  }
 
   /// This method is called after visiting the children of a decl. If it returns
   /// false, the remaining traversal is terminated and returns failure.
-  virtual bool walkToDeclPost(Decl *D) { return true; }
+  virtual bool walkToDeclPost(Decl *D) {
+    endBalancedDeclVisit(D);
+    return true;
+  }
+
+  /// This method is called immediately before \c walkToDeclPre. It is guranteed
+  /// to be called in balance with its counterpart \c endBalancedDeclVisit even
+  /// if \c walkToDeclPre returns false.
+  virtual void beginBalancedDeclVisit(Decl *D){};
+
+  /// This method is called immediately after \c walkToDeclPost or after \c
+  /// walkToDeclPre if it returns false.
+  virtual void endBalancedDeclVisit(Decl *D){};
```

Comment:
I don't _love_ the addition of these two new methods, however I needed a place to setup the `ContainerStack` before `SemaAnnotator::walkToDeclPre` begins walking. I deliberately haven't alluded to their intended use in their naming, as the `ContainerStack` is solely a concern of `IndexSwiftASTWalker`. I also needed their calls to be balanced, unlike `walkToDeclPre/Post`. I noticed some existing logic and commits dealing with other call balance issues, so it seems to me that the nature of `ASTWalker` to not always balance `walkToDeclPre/Post` calls is a little problematic. Two possible solutions to this would be to move these new methods down into `ASTWalker`, or modify `ASTWalker` to always call `walkToDeclPost`, but indicate the return value of `walkToDeclPre` with a new argument. I think both of these approaches would be far too invasive for my first ever contribution, and certainly outside the scope of this PR.

Reply:
Yeah, I agree this is unfortunate.  I kind of like the idea of changing ASTWalker and SourceEntityWalker to always call the Post function, but it could be a fairly invasive change - I'm not sure how deeply that assumption is baked into the existing subclasses - so I'm not sure one way or the other.

@bnbarham or @rintaro do you have any suggestions or thoughts here?

Reply:
(Without closely looking into the patch)
In general, _not_ calling `walkToDeclPost` if `walkToDeclPre` returned false sounds like the correct behavior to me. It's like calling `button.push()` doesn't/shouldn't invoke a `onPush` event in UI programming. Since the user (i.e. the type implementing `walkToDeclPre`) have the complete control of whether to visit the children, it should be easy to invoke the balanced operation manually.

```cpp
bool walkToDeclPre(Decl *D) {
  callBalancedBegin(D);

  if (condition) {
    callBalancedEnd(D);

    // don't visit children.
    return false;
  }

  // visit children.
  return true;
}

bool walkToDeclPost(Decl *D) {
  callBalancedEnd(D);
}
```

Reply:
Unfortunately I can't push all this down into `IndexSwiftASTWalker` because `SemaAnnotator::walkToDeclPre` calls `handleCustomAttributes` (plus some other stuff) prior to calling `SourceEntityWalker::walkToDeclPre`. I need the container setup before `handleCustomAttributes` is called so that the var also contains property wrapper references.

Reply:
So first off - I only just noticed that you had updated your forum post, sorry for missing these! I was expecting notifications, but I forgot to set "watching" on the topic. Feel free to @ me/ping on the Swift Slack if I'm unresponsive.

Looks like we actually want the contained by relationship for references in:
  * Declaration type
  * Initializers
  * Custom attributes

I missed that `handleCustomAttributes` was being called after the `beginBalancesDeclVisit` call when first reviewing. The idea of the SourceEntityWalker is to walk in source order, so it's a little weird to have a decl visit method that would be out of source order (eg. the VarDecl and then the custom attribute in this case).

You mention in the thread the possibility of new `walkToContainingDeclPre/Post` methods. I'm thinking something like that is probably the go, along with custom walks for the PDB and pattern in SemaAnnotator (which would call them). Perhaps `visitContainerPre/Post`? It would have to be called multiple times to handle eg.
```
var (a, b): (Int, String) = (foo(), bar())
```
Where the visit order would be `a`, `b`, `Int`, `String`, `foo`, `bar`. Ie. `pre(a) -> Int -> post(a) -> pre(b) -> String -> post(b) -> pre(a) -> foo -> post(a) -> pre(b) -> bar -> post(b)`

@benlangmuir any thoughts?

Reply:
> So first off - I only just noticed that you had updated your forum post, sorry for missing these! I was expecting notifications, but I forgot to set "watching" on the topic. Feel free to @ me/ping on the Swift Slack if I'm unresponsive.

No worries at all! I wasn't aware of there being a public Swift Slack, or is that some internal Apple thing?

> I missed that handleCustomAttributes was being called after the beginBalancesDeclVisit call when first reviewing. The idea of the SourceEntityWalker is to walk in source order, so it's a little weird to have a decl visit method that would be out of source order (eg. the VarDecl and then the custom attribute in this case).

I agree that it's a little weird, though I can't really think of a great way to avoid it. Perhaps `beginBalancedDeclVisit` could be renamed to indicate it's not necessarily in source order?

> You mention in the thread the possibility of new walkToContainingDeclPre/Post methods.

Yes that was my initial plan, however I found that messing with the existing walking got messy very quickly. That's why I went with the `begin/endBalancedDeclVisit` approach, it allows me to just save the container decl on a stack, and leave all the existing walking logic the same.

Note that I deliberately didn't end up going with `walkToContainingDeclPre/Post` as the concept of a container is solely a concern of `IndexSwiftASTWalker`. Implementing a `walkToContainingDeclPre/Post` in `SemaAnnotator` implies that it's only only to be called for container typed, which means I'd need to perform duplicate type casts in both `SemaAnnotator` and `IndexSwiftASTWalker`.

On the destructuring front, I've actually got something working that doesn't involve a custom walk. In a nutshell, it identifies important "singposts" (Expr, TypeRepr) in the pattern binding, along with the `VarDecl` they're associated with. Then, in `walkToExprPre/walkToTypeReprPre` it checks if the `Expr/TypeRepr` is a known "signpost", and if so, it activates the associated `VarDecl`. Then when the normal walk comes to request the container, it just returns whatever the current active `VarDecl` is. I should have it ready for review by tomorrow.

Reply:
> No worries at all! I wasn't aware of there being a public Swift Slack, or is that some internal Apple thing?

Ah yep, sorry about that! @/DM on Swift Forums would be ideal :)